crdclient/crd_client.go

```diff
@@ -0,0 +1,88 @@
+package crdclient
+
+import (
+	"context"
+
+	"github.com/cenk/backoff"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/giantswarm/operatorkit/crd"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	apiextensionsclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type Config struct {
+	// Dependencies.
+
+	K8sExtClient apiextensionsclient.Interface
+	Logger       micrologger.Logger
+}
+
+type CRDClient struct {
+	k8sExtClient apiextensionsclient.Interface
+	logger       micrologger.Logger
+}
+
+func New(config Config) (*CRDClient, error) {
+	// Dependencies.
+	if config.K8sExtClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.K8sExtClient must not be empty")
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.Logger must not be empty")
+	}
+
+	crdClient := &CRDClient{
+		k8sExtClient: config.K8sExtClient,
+		logger:       config.Logger,
+	}
+
+	return crdClient, nil
+}
+
+// Ensure ensures CRD exists, it is active (aka. established) and does not have
+// conflicting names.
+func (c *CRDClient) Ensure(ctx context.Context, customResource *crd.CRD, backOff backoff.BackOff) error {
+	_, err := c.k8sExtClient.ApiextensionsV1beta1().CustomResourceDefinitions().Create(customResource.NewResource())
+	if errors.IsAlreadyExists(err) {
+		// Fall trough. We need to check CRD status.
```

Comment:
This is one-line, but important change. `manifest.Status.Conditions` are checked even if CRD already exists.

Otherwise it's all copy/paste + adding Config and New.

Reply:
@xh3b4sd did you notice that? Does that make sense?

Reply:
That works for me. It is a bit like an reconciliation approach to ensure the current state matches the actual desired state. All good. <3